### PR TITLE
Change source endpoint with source-type and add format field to source interface

### PR DESCRIPTION
**Pull request by hitarthpattani**
### Description (*)

Change source upload endpoint and provide a way to define source type within the request URL. Define new format field to database and store format field object as JSON string in the database table.

### Fixed Issues (if relevant)

1. magento/async-import#75: Change Source Upload endpoints

### Manual testing scenarios (*)

1.  URL (POST request): [DOMAIN]/rest/V1/import/source/csv
2. Body
```
{
	"source": {
		"import_data": "var/catalog_product.csv",
		"import_type": "local_path",
		"format": {
			"csv_separator": ",",
			"csv_enclosure": "\"",
			"csv_delimiter": ",",
			"multiple_value_separator": "|",
			"mapping": [
				{
					"source_attribute": "string",
					"destination_attribute": "string",
					"processing_rules": "string",
					"taxonomy": "string",
					"values_mapping": [
						{
							"old_value": "value1",
							"new_value": "value2"
						}, {
							"old_value": "value3",
							"new_value": "value4"
						}
					]
				}, {
					"source_attribute": "string",
					"destination_attribute": "string",
					"processing_rules": "string",
					"taxonomy": "string"
				}
			]
		}
	}
}
```
3. Headers: Content-Type: application/json, Authorization: Bearer <admin_user_token>

Response should be:
```
{
	"source_id": 74,
	"status": "uploaded",
	"error": null,
	"source": {
		"source_id": 74,
		"source_type": "csv",
		"import_type": "local_path",
		"status": "uploaded",
		"import_data": "5c9b7620773cd.csv",
		"format": {
			"csv_separator": ",",
			"csv_enclosure": "\"",
			"csv_delimiter": ",",
			"multiple_value_separator": "|",
			"mapping": [
				{
					"source_attribute": "string",
					"destination_attribute": "string",
					"processing_rules": "string",
					"taxonomy": "string",
					"values_mapping": [
						{
							"old_value": "value1",
							"new_value": "value2"
						},
						{
							"old_value": "value3",
							"new_value": "value4"
						}
					]
				},
				{
					"source_attribute": "string",
					"destination_attribute": "string",
					"processing_rules": "string",
					"taxonomy": "string",
					"values_mapping": null
				}
			]
		},
		"created_at": "2019-03-27 13:09:52"
	}
}
```

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/ImportService/etc/db_schema.xml [hunk 1]
```diff
@@ -13,6 +13,7 @@
         <column xsi:type="varchar" name="source_type" length="255" comment="Type of imported sources"/>
         <column xsi:type="varchar" name="import_type" length="255" comment="Type of the import"/>
         <column xsi:type="varchar" name="import_data" length="255" comment="Import data"/>
+        <column xsi:type="text" name="format" comment="Source Format" />
```
[the remark this answers is not in the document]
Comment (hitarthpattani) [reply app/code/Magento/ImportService/etc/db_schema.xml]: I have changed field type from text to longtext.

**Comment by hitarthpattani** [comment]
Hello @nuzil 

I have resolved the issue on "Call to a member function getMapping() on null".

**Comment by nuzil** [comment]
Great. One more point
Currently you are saving NULL in case if no format.
I would propose to define in code default values for params: 

“csv_separator”: “,”,
           “csv_enclosure”: “\”“,
           “csv_delimiter”: “,”,
           “multiple_value_separator”: “|”,


And on save action, if they are not presented in request please set those values from defaults and save them into database

**Comment by hitarthpattani** [comment]
Hello @nuzil I have made mentioned changes of setting default values for format field in the source object.

File: app/code/Magento/ImportService/Api/Data/SourceFormatInterface.php [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+     */
+    public function getCsvSeparator();
+
+    /**
+     * Set CSV Separator
+     *
+     * @param string $csvSeparator
+     * @return $this
+     */
+    public function setCsvSeparator($csvSeparator);
```
Comment (nuzil) [inline app/code/Magento/ImportService/Api/Data/SourceFormatInterface.php]: Can you please use correct tupes definition for params and return values.
Here is a good example: 
https://github.com/magento/async-import/pull/66/files#diff-edfcdbe6108e5d99eef61e317ca74c80

So in this case: 
public function setCsvSeparator(string $csvSeparator): void;

Please refactor all Data interfaces in this PR

Comment (hitarthpattani) [reply app/code/Magento/ImportService/Api/Data/SourceFormatInterface.php]: Hello, @nuzil I have refectored all the data interface.

**Comment by nuzil** [comment]
Hi @hitarthpattani 
I just merged PR what related to using UUIDs and not source_ids 
https://github.com/magento/async-import/pull/53

Can you please back merge and align changes in your code?

File: app/code/Magento/ImportService/Api/Data/SourceInterface.php [hunk 3]
```diff
@@ -19,94 +20,113 @@ interface SourceInterface extends ExtensibleDataInterface
...
      */
-    public function setImportData($importData);
+    public function setImportData(string $importData): SourceInterface;
+
+    /**
+     * Retrieve Source Format
+     *
+     * @return \Magento\ImportService\Api\Data\SourceFormatInterface
+     */
+    public function getFormat(): SourceFormatInterface;
```
Comment (nuzil) [inline app/code/Magento/ImportService/Api/Data/SourceInterface.php]: This method also can return null.
Please align this in annotations and also in respective Models in return type

File: app/code/Magento/ImportService/Model/Source.php [hunk 4]
```diff
@@ -40,95 +97,111 @@ public function getIdentities()
     /**
      * @inheritDoc
      */
-    public function getUuid()
+    public function getUuid(): string
```
Comment (nuzil) [inline app/code/Magento/ImportService/Model/Source.php]: This method also can return null.
Please align this in annotations and also in respective Models in return type

File: app/code/Magento/ImportService/Api/Data/SourceInterface.php [hunk 5]
```diff
@@ -19,94 +20,113 @@ interface SourceInterface extends ExtensibleDataInterface
...
     const STATUS_COMPLETED = 'completed';
     const STATUS_FAILED = 'failed';
 
     /**
+     * Retrieve source uuid
+     *
      * @return string
      */
-    public function getUuid();
+    public function getUuid(): string;
```
Comment (nuzil) [inline app/code/Magento/ImportService/Api/Data/SourceInterface.php]: This method also can return null.
Please align this in annotations and also in respective Models in return type

Comment (hitarthpattani) [reply app/code/Magento/ImportService/Api/Data/SourceInterface.php]: I have corrected the code.

File: app/code/Magento/ImportService/Model/Source.php (hunk 4 above)
Comment (hitarthpattani) [reply app/code/Magento/ImportService/Model/Source.php]: I have corrected the code.

File: app/code/Magento/ImportService/Api/Data/SourceInterface.php (hunk 3 above)
Comment (hitarthpattani) [reply app/code/Magento/ImportService/Api/Data/SourceInterface.php]: I have corrected the code.

File: app/code/Magento/ImportService/etc/webapi.xml [hunk 6]
```diff
@@ -7,7 +7,7 @@
 -->
 <routes xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Webapi:etc/webapi.xsd">
-    <route url="/V1/import/source" method="POST">
+    <route url="/V1/import/source/:sourceType" method="POST">
```
Comment (nuzil) [inline app/code/Magento/ImportService/etc/webapi.xml]: Hi @hitarthpattani 
Sorry I didnt noticed that before, but idea was to create fixed endpoint for each type. Which means that endpoint have to be fixed and looks like 
/V1/import/source/csv
sourceType - is not a variable at the point, its part of endpoint.
Reason for that, that XML format for example have another format fields which mean we have to be able to use another parameters for parsing (we dont need csv_* fields for example).  

Can you please fix endpoint and also change main interface name from SourceUploadInterface to SourceUploadCsvInterface

Comment (nuzil) [reply app/code/Magento/ImportService/etc/webapi.xml]: I would also give more detailed names for interfaces, like SourceCsv and SourceCsvFormatInterface and all relared

**Pull request merged**